[BUG] module compile error on latest kernel 6.14.0-rc4

**Problem Description**
Building of dkms module fails on latest linux kernel 6.14.0-rc4. The compile error is stated in make.log below.

This is because of this patchset: [[PATCH v4 00/19] Hide platform_profile_handler from consumers](https://lore.kernel.org/all/20250116002721.75592-1-kuurtb@gmail.com/#r).

**Model and Debug Info**

Content of dkms make.log:


```
DKMS make.log for legion-laptop-0.0.19+git2025-02-20.43f2ad0e for kernel 6.14.0-5-generic (x86_64)
Fri Feb 28 14:20:21 CET 2025
make: Entering directory '/usr/src/linux-headers-6.14.0-5-generic'
make[1]: Entering directory '/var/lib/dkms/legion-laptop/0.0.19+git2025-02-20.43f2ad0e/build'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: x86_64-linux-gnu-gcc-14 (Ubuntu 14.2.0-17ubuntu3) 14.2.0
  You are using:           gcc-14 (Ubuntu 14.2.0-17ubuntu3) 14.2.0
  CC [M]  legion-laptop.o
legion-laptop.c:2411:41: error: field ‘platform_profile_handler’ has incomplete type
 2411 |         struct platform_profile_handler platform_profile_handler;
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~
legion-laptop.c: In function ‘legion_platform_profile_notify’:
legion-laptop.c:4888:9: error: too few arguments to function ‘platform_profile_notify’
 4888 |         platform_profile_notify();
      |         ^~~~~~~~~~~~~~~~~~~~~~~
In file included from legion-laptop.c:74:
/usr/src/linux-headers-6.14.0-5-generic/include/linux/platform_profile.h:55:6: note: declared here
   55 | void platform_profile_notify(struct device *dev);
      |      ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/src/linux-headers-6.14.0-5-generic/include/linux/bits.h:22,
                 from /usr/src/linux-headers-6.14.0-5-generic/include/linux/ioport.h:13,
                 from /usr/src/linux-headers-6.14.0-5-generic/include/linux/acpi.h:12,
                 from legion-laptop.c:62:
legion-laptop.c: In function ‘legion_platform_profile_get’:
/usr/src/linux-headers-6.14.0-5-generic/include/linux/compiler_types.h:483:27: error: expression in static assertion is not an integer
  483 | #define __same_type(a, b) __builtin_types_compatible_p(typeof(a), typeof(b))
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/linux-headers-6.14.0-5-generic/include/linux/build_bug.h:78:56: note: in definition of macro ‘__static_assert’
   78 | #define __static_assert(expr, msg, ...) _Static_assert(expr, msg)
      |                                                        ^~~~
/usr/src/linux-headers-6.14.0-5-generic/include/linux/container_of.h:20:9: note: in expansion of macro ‘static_assert’
   20 |         static_assert(__same_type(*(ptr), ((type *)0)->member) ||       \
      |         ^~~~~~~~~~~~~
/usr/src/linux-headers-6.14.0-5-generic/include/linux/container_of.h:20:23: note: in expansion of macro ‘__same_type’
   20 |         static_assert(__same_type(*(ptr), ((type *)0)->member) ||       \
      |                       ^~~~~~~~~~~
legion-laptop.c:4897:16: note: in expansion of macro ‘container_of’
 4897 |         priv = container_of(pprof, struct legion_private,
      |                ^~~~~~~~~~~~
legion-laptop.c: In function ‘legion_platform_profile_set’:
/usr/src/linux-headers-6.14.0-5-generic/include/linux/compiler_types.h:483:27: error: expression in static assertion is not an integer
  483 | #define __same_type(a, b) __builtin_types_compatible_p(typeof(a), typeof(b))
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/linux-headers-6.14.0-5-generic/include/linux/build_bug.h:78:56: note: in definition of macro ‘__static_assert’
   78 | #define __static_assert(expr, msg, ...) _Static_assert(expr, msg)
      |                                                        ^~~~
/usr/src/linux-headers-6.14.0-5-generic/include/linux/container_of.h:20:9: note: in expansion of macro ‘static_assert’
   20 |         static_assert(__same_type(*(ptr), ((type *)0)->member) ||       \
      |         ^~~~~~~~~~~~~
/usr/src/linux-headers-6.14.0-5-generic/include/linux/container_of.h:20:23: note: in expansion of macro ‘__same_type’
   20 |         static_assert(__same_type(*(ptr), ((type *)0)->member) ||       \
      |                       ^~~~~~~~~~~
legion-laptop.c:4926:16: note: in expansion of macro ‘container_of’
 4926 |         priv = container_of(pprof, struct legion_private,
      |                ^~~~~~~~~~~~
legion-laptop.c: In function ‘legion_platform_profile_init’:
legion-laptop.c:4974:15: error: too few arguments to function ‘platform_profile_register’
 4974 |         err = platform_profile_register(&priv->platform_profile_handler);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/linux-headers-6.14.0-5-generic/include/linux/platform_profile.h:47:16: note: declared here
   47 | struct device *platform_profile_register(struct device *dev, const char *name,
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~
legion-laptop.c: In function ‘legion_platform_profile_exit’:
legion-laptop.c:4988:9: error: too few arguments to function ‘platform_profile_remove’
 4988 |         platform_profile_remove();
      |         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/src/linux-headers-6.14.0-5-generic/include/linux/platform_profile.h:50:5: note: declared here
   50 | int platform_profile_remove(struct device *dev);
      |     ^~~~~~~~~~~~~~~~~~~~~~~
legion-laptop.c: At top level:
legion-laptop.c:1827:16: warning: ‘ecram_memoryio_write’ defined but not used [-Wunused-function]
 1827 | static ssize_t ecram_memoryio_write(const struct ecram_memoryio *ec_memoryio,
      |                ^~~~~~~~~~~~~~~~~~~~
make[3]: *** [/usr/src/linux-headers-6.14.0-5-generic/scripts/Makefile.build:207: legion-laptop.o] Error 1
make[2]: *** [/usr/src/linux-headers-6.14.0-5-generic/Makefile:1994: .] Error 2
make[1]: *** [/usr/src/linux-headers-6.14.0-5-generic/Makefile:251: __sub-make] Error 2
make[1]: Leaving directory '/var/lib/dkms/legion-laptop/0.0.19+git2025-02-20.43f2ad0e/build'
make: *** [Makefile:251: __sub-make] Error 2
make: Leaving directory '/usr/src/linux-headers-6.14.0-5-generic'
```

--- a/kernel_module/legion-laptop.c
+++ b/kernel_module/legion-laptop.c
@@ -2408,7 +2408,11 @@ struct legion_private {
 	//interfaces
 	struct dentry *debugfs_dir;
 	struct device *hwmon_dev;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+	struct device *ppdev;
+#else
 	struct platform_profile_handler platform_profile_handler;
+#endif
 
 	struct light kbd_bl;
 	struct light ylogo_light;
@@ -4637,7 +4641,11 @@ static ssize_t powermode_show(struct device *dev, struct device_attribute *attr,
 	return sysfs_emit(buf, "%d\n", power_mode);
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+static void legion_platform_profile_notify(struct device *dev);
+#else
 static void legion_platform_profile_notify(void);
+#endif
 
 static ssize_t powermode_store(struct device *dev,
 			       struct device_attribute *attr, const char *buf,
@@ -4662,7 +4670,11 @@ static ssize_t powermode_store(struct device *dev,
 	// readback done after notifying returns correct value, otherwise
 	// the notified reader will read old value
 	msleep(500);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+	legion_platform_profile_notify(priv->ppdev);
+#else
 	legion_platform_profile_notify();
+#endif
 
 	return count;
 }
@@ -4780,7 +4792,11 @@ unlock:
 	// problem: we get an event just before the powermode change (from the key?),
 	// so if we notify too early, it will read the old power mode/platform profile
 	msleep(500);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+	legion_platform_profile_notify(priv->ppdev);
+#else
 	legion_platform_profile_notify();
+#endif
 }
 
 static int legion_wmi_probe(struct wmi_device *wdev, const void *context)
@@ -4880,22 +4896,38 @@ static void legion_wmi_exit(void)
 /* Platform profile               */
 /* ============================   */
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+static void legion_platform_profile_notify(struct device *dev)
+#else
 static void legion_platform_profile_notify(void)
+#endif
 {
 	if (!enable_platformprofile)
 		pr_info("Skipping platform_profile_notify because enable_platformprofile is false\n");
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+	platform_profile_notify(dev);
+#else
 	platform_profile_notify();
+#endif
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+static int legion_platform_profile_get(struct device *dev,
+				       enum platform_profile_option *profile)
+#else
 static int legion_platform_profile_get(struct platform_profile_handler *pprof,
 				       enum platform_profile_option *profile)
+#endif
 {
 	int powermode;
 	struct legion_private *priv;
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+	priv = dev_get_drvdata(dev);
+#else
 	priv = container_of(pprof, struct legion_private,
 			    platform_profile_handler);
+#endif
 	read_powermode(priv, &powermode);
 
 	switch (powermode) {
@@ -4917,14 +4949,23 @@ static int legion_platform_profile_get(struct platform_profile_handler *pprof,
 	return 0;
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+static int legion_platform_profile_set(struct device *dev,
+				       enum platform_profile_option profile)
+#else
 static int legion_platform_profile_set(struct platform_profile_handler *pprof,
 				       enum platform_profile_option profile)
+#endif
 {
 	int powermode;
 	struct legion_private *priv;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+	priv = dev_get_drvdata(dev);
+#else
 	priv = container_of(pprof, struct legion_private,
 			    platform_profile_handler);
+#endif
 
 	switch (profile) {
 	case PLATFORM_PROFILE_BALANCED:
@@ -4946,8 +4987,34 @@ static int legion_platform_profile_set(struct platform_profile_handler *pprof,
 	return write_powermode(priv, powermode);
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+static bool conf_has_custom_powermode;
+static enum access_method conf_access_method_powermode;
+
+static int legion_platform_profile_probe(void *drvdata, unsigned long *choices)
+{
+	set_bit(PLATFORM_PROFILE_QUIET, choices);
+	set_bit(PLATFORM_PROFILE_BALANCED, choices);
+	set_bit(PLATFORM_PROFILE_PERFORMANCE, choices);
+	if (conf_has_custom_powermode && conf_access_method_powermode == ACCESS_METHOD_WMI) {
+		set_bit(PLATFORM_PROFILE_BALANCED_PERFORMANCE, choices);
+	}
+
+	return 0;
+}
+
+static const struct platform_profile_ops legion_platform_profile_ops = {
+	.probe = legion_platform_profile_probe,
+	.profile_get = legion_platform_profile_get,
+	.profile_set = legion_platform_profile_set,
+};
+#endif
+
 static int legion_platform_profile_init(struct legion_private *priv)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+	struct device *dev = &priv->platform_device->dev;
+#endif
 	int err;
 
 	if (!enable_platformprofile) {
@@ -4955,6 +5022,10 @@ static int legion_platform_profile_init(struct legion_private *priv)
 		return 0;
 	}
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+	conf_has_custom_powermode = priv->conf->has_custom_powermode;
+	conf_access_method_powermode = priv->conf->access_method_powermode;
+#else
 	priv->platform_profile_handler.profile_get =
 		legion_platform_profile_get;
 	priv->platform_profile_handler.profile_set =
@@ -4970,10 +5041,17 @@ static int legion_platform_profile_init(struct legion_private *priv)
 		set_bit(PLATFORM_PROFILE_BALANCED_PERFORMANCE,
 			priv->platform_profile_handler.choices);
 	}
+#endif
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0)
+	priv->ppdev = devm_platform_profile_register(dev, "lenovo-legion", priv, &legion_platform_profile_ops);
+	if (IS_ERR(priv->ppdev))
+		return PTR_ERR(priv->ppdev);
+#else
 	err = platform_profile_register(&priv->platform_profile_handler);
 	if (err)
 		return err;
+#endif
 
 	return 0;
 }
@@ -4985,7 +5063,9 @@ static void legion_platform_profile_exit(struct legion_private *priv)
 		return;
 	}
 	pr_info("Unloading legion platform profile\n");
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 14, 0)
 	platform_profile_remove();
+#endif
 	pr_info("Unloading legion platform profile done\n");
 }
 

